concordium-smart-contract-testing/src/invocation/impls.rs

```diff
@@ -0,0 +1,1468 @@
+                                from_interpreter_energy(InterpreterEnergy::from(remaining_energy));
+                            remaining_energy = remaining_energy.saturating_sub(
+                                constants::CONTRACT_INSTANCE_QUERY_ACCOUNT_BALANCE_COST,
+                            );
+
+                            let resume_res = run_interpreter(remaining_energy, |energy| {
+                                v1::resume_receive(
+                                    config,
+                                    response,
+                                    energy,
+                                    &mut self.state,
+                                    false, // State never changes on queries.
+                                    v1::trie::Loader::new(&[][..]),
+                                )
+                            });
+
+                            self.process(resume_res)
+                        }
+                        v1::Interrupt::QueryContractBalance { address } => {
+                            let response = match self.invocation_handler.contract_balance(address) {
+                                None => v1::InvokeResponse::Failure {
+                                    kind: v1::InvokeFailure::NonExistentContract,
+                                },
+                                Some(bal) => v1::InvokeResponse::Success {
+                                    // Balance of contract querying. Won't change. Notice the
+                                    // `self.address`.
+                                    new_balance: self
+                                        .invocation_handler
+                                        .contract_balance_unchecked(self.address),
+                                    data:        Some(to_bytes(&bal)),
+                                },
+                            };
+
+                            let mut remaining_energy =
+                                from_interpreter_energy(InterpreterEnergy::from(remaining_energy));
+                            remaining_energy = remaining_energy.saturating_sub(
+                                constants::CONTRACT_INSTANCE_QUERY_CONTRACT_BALANCE_COST,
+                            );
+
+                            let resume_res = run_interpreter(remaining_energy, |energy| {
+                                v1::resume_receive(
+                                    config,
+                                    response,
+                                    energy,
+                                    &mut self.state,
+                                    false, // State never changes on queries.
+                                    v1::trie::Loader::new(&[][..]),
+                                )
+                            });
+
+                            self.process(resume_res)
+                        }
+                        v1::Interrupt::QueryExchangeRates => {
+                            let exchange_rates = (
```

Comment:
The serialization seems correct in this case, but maybe we should move the `ExchangeRates` type from `concordium-std` and reuse it here. Otherwise, it is very subtle why this is correct. (the order and whether it is endianess)

Reply:
Yes, please do that.